### DISCUSSION TITLE
Blazor Web Assembly

### DISCUSSION THREAD

**Discussion by seanjmiller** [melanchall/drywetmidi#221] [General]
Trying to port my Windows Forms app to Blazor Web Assembly. Not surprised there are issues with Native code so I tried switching to the Nativeless and OutputDevice is not a member. I did see the documentation on the Nativeless version clearly stating this. Is there anything I can do to Playback a midi file? Thanks
Here is the last few lines of my code for reference:

`        using (var outputDevice = OutputDevice.GetByName("Microsoft GS Wavetable Synth"))
        using (var playback = midiFile.GetPlayback(outputDevice))
        {
            playback.Play();
        }
`

**Comment by melanchall** [comment]
Can you please provide details of what errors with native API do you have?

It would be great if you provide me your entire solution so I can reproduce the issue.

**Reply by seanjmiller** [reply]
So the GetByName throws an exception and here are the lines that would interest you:

crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Melanchall_DryWetMidi_Native32
System.DllNotFoundException: Melanchall_DryWetMidi_Native32
   at Melanchall.DryWetMidi.Multimedia.CommonApi32.Api_GetApiType()
   at Melanchall.DryWetMidi.Multimedia.MidiDevicesSession.GetSessionHandle()
   at Melanchall.DryWetMidi.Multimedia.MidiDevice.EnsureSessionIsCreated()
   at Melanchall.DryWetMidi.Multimedia.OutputDevice.GetByName(String name)

The exact same code runs under my Windows Form project. If you need me to I can create a Blazor Web Assembly with sample code that you have provided to get started and I am sure the same result will happen.

**Reply by melanchall** [reply]
> If you need me to I can create a Blazor Web Assembly with sample code that you have provided to get started and I am sure the same result will happen.

Yes, please. Send me the solution so I can get exactly the same error. `DllNotFoundException` just means native binaries are not in a right place.

**Reply by seanjmiller** [reply]
Ok I created a Blazor Web Assembly right out of the Project Templates and added Melanchall.DryWetMidi and added your sample code to the Counter Page. Just click the 'Click me' button which will run the code and generate the exception. I added you as a collaborator to this project.

**Reply by melanchall** [reply]
Thanks! I'll be able to look into it a week later. I'm on vacation right now without access to my computer.

**Reply by seanjmiller** [reply]
No problem thanks for the support!

**Reply by melanchall** [reply]
@seanjmiller Well, I've cloned the repo and successfully reproduced the error. I made some research and the conclusion is: we need to have native code in WebAssembly (wasm) file, but right now it's not possible to compile existing C code into wasm since it's not cross-platform. So creating a wasm file for DryWetMIDI requires a huge work, so at the moment unfortunately you can't use the library in Blazor.
